src/typechecker.erl

```diff
@@ -2302,7 +2302,19 @@ do_type_check_expr_in(Env, ResTy, {call, P, Name, Args}) ->
 do_type_check_expr_in(Env, ResTy, {lc, P, Expr, Qualifiers}) ->
     type_check_comprehension_in(Env, ResTy, lc, Expr, P, Qualifiers);
 do_type_check_expr_in(Env, ResTy, {bc, P, Expr, Qualifiers}) ->
-    type_check_comprehension_in(Env, ResTy, bc, Expr, P, Qualifiers);
+    {Res, Acc} = lists:mapfoldl(
+        fun(FlatTy, Acc) ->
+            case catch {type_check_comprehension_in(Env, FlatTy, bc, Expr, P, Qualifiers), Acc} of
+                {{#{}, C1}, {#{}, C2}}                -> {ok, {#{}, constraints:combine(C1, C2)}};
+                {{#{}, _} = C, {type_error, _, _, _}} -> {ok, C};
+                Err when element(1, Err)=:=type_error -> {Err,Acc}
+            end
+        end, {type_error, bc, P, ResTy}, flatten_type(ResTy, Env)),
+    case {lists:any(fun(El) -> El =:= ok end, Res), proplists:lookup(type_error,Res)} of
+        {true, _} -> Acc;
+        {_, none} -> Acc;
+        {_,Err} -> throw(Err)
+    end;
```

Comment:
The tuple `{type_error, _, _, _}` is not enough for a mismatch. For example, if a type such as `binary() | nonexisting_module:foo()` is expected (which easily happens if the code path to deps isn't available), you will get a crash here. In general, all exceptions of class `throw` can happen, so use try-catch and some more exact logic. (Take a look at `handle_type_error` and all the throws in this module to see what it can be. More can also be added in the future.)

Btw, this is way different to how the same thing is done for list comprehensions. Take a look at `expect_list_type/1`. It isn't even using exceptions, but filters the type union on a different level. I'm not saying it has to be done the same way for bc, but consider the differences. (Personally, I'd like to replace these with loop over the union on a higher level, in a clause for `do_type_check_expr_in(Env, {type, _, union, Tys}, Expr)`, but that's a different PR and discussion.)

Don't use flatten_type/2. It's an unintuitive function because it turns a type into a list of types, which isn't clear from its name. It's intended as a helper for flatten_unions/2, which is a helper for normalize/2, only. (Yes, they should be private in another module, I agree.) Pattern match on `{type, union, _, _}` instead.

I think this is somewhat hard to read as well. Consider breaking it up into multiple functions.

Reply:
Ah, there's one way to do all the stuff I was doing on only three lines of codes, that `type_check_union_in` that was used for the list comprehension case. What do you think now?